### ISSUE TITLE
Cannot achieve >=0.8 accuracy

### ISSUE THREAD

**Issue by nsamko** [mikeizbicki/cmc-csci181#19]
Vik, Nik and I are trying to find the most optimal hyperparameters that would satisfy

`loss_nextchars` < 1.5 (for both models)
`accuracy` >= 0.8 (only for the model without the `--conditional_model` flag)

However, we still cannot achieve the required accuracy. Since we have to use an optimizer other than SGD, we tried Adam and LSTM as well as Adam and RNN. The best we could achieve without `--conditional_flag` was by using

`python3 hw7.py --train --model=rnn --learning_rate=1e-3 --batch_size=10 --gradient_clipping --optimizer=adam --samples=100000`

Here are our tensorboard links for LTSM and Adam:
conditional: https://tensorboard.dev/experiment/vpPRHY3xT6CQRSYPHLLhWA/
unconditional: https://tensorboard.dev/experiment/Dk8AVEP1Rla1XC1k7iVkWQ/

Any suggestions on how to improve the results?

**Comment by mikeizbicki** [comment]
It looks to me like your blue model would achieve >= 0.8 accuracy if you let it run for the full 100k samples instead of stopping it after 30k.  It's already at the cusp of reaching that accuracy.

Also, don't forget that you're required to warm start the model twice with lower learning rates.  You should see significant jumps in accuracy when you do this, and I would expect that your blue model will get an accuracy of about 0.95 after doing this.

**Comment by zhh1997zhh** [comment]
@mikeizbicki 
Hi professor, I modified my command line so that the initial learning rate of my model is 1e-3, however, it seems that my code is still unable to get accuracy above 0.8 (from 1e-3 to 1e-5). May I know if there is a way to fix the problem? (For some reasons, lowering the learning rate from 1e-4 to 1e-5 doesn't make much difference)

My command line example (without conditional model flag): 
```
$ python3 names.py --train --model=lstm --gradient_clipping --optimizer=adam --samples=100000 --learning_rate=1e-5 --warm_start=log/lstm4
```

My other parameters (I just use their default values):
```
parser_model.add_argument('--hidden_layer_size',type=int,default=128)
parser_model.add_argument('--num_layers',type=int,default=1)
parser_opt.add_argument('--batch_size',type=int,default=1)
parser_opt.add_argument('--momentum',type=float,default=0.9)
parser_opt.add_argument('--weight_decay',type=float,default=1e-4)
parser_opt.add_argument('--input_length',type=int,default=20)
```

My tensorboard link:
https://tensorboard.dev/experiment/nZUnE4FwRru7OJnoDKFLZg/

**Comment by mikeizbicki** [comment]
@zhh1997zhh  Ahh... I think I see the problem now.

**short answer:** Your batch size of 1 is very small.  I think if you change the batch size to 10 than this will fix the problem.

**long answer:** The performance of your model depends on the total number of data points that your model has looked at, and there are two hyperparameters that control this: `--batch_size` and `--num_samples`.  With a `--batch_size=1` and `--num_samples=100000`, you are looking at a total of 100,000 data points.  When I ran the model with `--batch_size=10` and `--num_samples=100000`, I was looking at 1,000,000 data points, 10x as many, and that's why my lstm/adam model performed better.  If you don't want to increase your batch size, you could alternatively make your model look at more data points by increasing the `--num_samples` to `1000000`, and that would have a similar effect.

If you look at your tensorboard link, the orange line (lstm3) never actually converges to its final steady-state value.  You can still see that the accuracy is still increasing, and the loss values are still decreasing after 100k iterations.  This tells us that the model needs to be trained for longer with the current hyperparameter values.
